lib/rucio/core/replica.py

```diff
@@ -2478,15 +2658,21 @@ def get_source_replicas(scope, name, source_rses=None, *, session: "Session"):
     :returns:              List of SQLAlchemy Replica Objects
     """
 
-    query = session.query(models.RSEFileAssociation.rse_id).filter_by(scope=scope, name=name).filter(models.RSEFileAssociation.state == ReplicaState.AVAILABLE)
+    stmt = select(
+        models.RSEFileAssociation.rse_id
+    ).where(
+        and_(models.RSEFileAssociation.scope == scope,
+             models.RSEFileAssociation.name == name,
+             models.RSEFileAssociation.state == ReplicaState.AVAILABLE)
+    )
     if source_rses:
         if len(source_rses) < 10:
             rse_clause = []
             for rse_id in source_rses:
                 rse_clause.append(models.RSEFileAssociation.rse_id == rse_id)
             if rse_clause:
-                query = query.filter(or_(*rse_clause))
-    return [a[0] for a in query.all()]
+                stmt = stmt.where(or_(*rse_clause))
+    return session.execute(stmt).scalars().all()
```

Comment:
Docs say `.all()` returns a sequence object, not a list. Does this mess with typing anywhere?

Reply:
I don't think so? I believe we only use `__iter__` on the results of this fn (`get_and_lock_file_replicas`) `Sequence` implements this. cc: typeguy @rdimaio

Reply:
`Sequence` implements `__iter__` as a mix-in: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes (and generally I think we would have seen errors in the CI if there was any issue)

Looking into the SQLA code, it seems like `ScalarResult.all()`  actually returns a `list` anyway:

`ScalarResult.all()` is typed as returning a `Sequence`: https://github.com/sqlalchemy/sqlalchemy/blob/8f47774de91b18e8b9a21d39662bffd239785b4e/lib/sqlalchemy/engine/result.py#L1793-L1801

But all it does is return `self._allrows`, which is inherited from `ResultInternal`, which specifically returns a `list`: https://github.com/sqlalchemy/sqlalchemy/blob/8f47774de91b18e8b9a21d39662bffd239785b4e/lib/sqlalchemy/engine/result.py#L545, and is type annotated as such

I'm not sure why they used the more generic `Sequence` type annotation for `ScalarResult.all()`. I'll open a PR to them to address this and see if it's a mistake or if there's a reason behind it

Reply:
PR for reference: https://github.com/sqlalchemy/sqlalchemy/pull/11584

Reply:
There are reasons behind the `Sequence` type annotation, see PR for context. No issue on our side though - when we'll type hint `get_source_replicas` we can either annotate it to return a `Sequence` or `list` (in the latter case we'll have to use `typing.cast(list,...)` or `list(...)`

Reply:
Thanks, @rdimaio. I'll leave it as is for now then, the docstring says it returns a `list`, and regardless of what the SQLA docs say, `.all()` does actually return a `list` here